Fatal error when converting specific Terraform code to Python

### What happened?

When running convert of specific Terraform code snippet described below to Python, `pulumi convert` returns following error: `panic: runtime error: index out of range [1] with length 1`

### Example

Minimal example of Terraform code to generate error:
```
resource "random_password" "user" {
  for_each = toset(["bob", "john", "carl"])
}

resource "aws_secretsmanager_secret" "db_users" {
  for_each                = toset(["bob", "john", "carl"])
  name                    = "platform/dev/mongoatlas/user/${each.key}"
}

    
resource "aws_secretsmanager_secret_version" "db_users" {
  for_each = toset(["bob", "john", "carl"])

  secret_id = aws_secretsmanager_secret.db_users[each.key].id
  secret_string = jsonencode(
    {
      username = each.key
      password = random_password.user[each.key].result
    }
  )
}
```

**if comment out line `password = random_password.user[each.key].result` then convertation happen successfully**

### Output of `pulumi about`

```CLI          
Version      3.85.0
Go Version   go1.21.1
Go Compiler  gc

Host     
OS       arch
Version  "rolling"
Arch     x86_64

Pulumi locates its logs in /tmp by default
```

### Additional context

Convert command and output:

```
$ pulumi convert --from terraform --language python --out pulumi
Converting from terraform...
warning: /home/andy/Work/SignantHealth/idp-module-mongoatlas-mongodb/test/main.tf:2,14-44: Function not yet implemented; Function toset not yet implemented
warning: /home/andy/Work/SignantHealth/idp-module-mongoatlas-mongodb/test/main.tf:9,29-59: Function not yet implemented; Function toset not yet implemented
warning: /home/andy/Work/SignantHealth/idp-module-mongoatlas-mongodb/test/main.tf:16,14-44: Function not yet implemented; Function toset not yet implemented
Converting to python...
panic: runtime error: index out of range [1] with length 1

goroutine 31 [running]:

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).rewriteTraversal(0xc0008e3c80, {0xc000690f60, 0x1, 0x1}, {0x24a1780, 0xc000a32cc0}, {0xc00402fd80, 0x1, 0x1fb7f08?})

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program_quotes.go:71 +0x1305

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).rewriteQuotes.func1({0x24a1900?, 0xc00071ba40?})

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program_quotes.go:271 +0x70

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitRelativeTraversal(0xc00071ba40, 0xc00071ba40?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:201 +0x10d

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x24a1900?, 0xc00071ba40?}, 0xc00840dca0?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:331 +0x1c6

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitExpressions({0xc003f2c600, 0x2, 0x2}, 0xc00840dca0?, 0x24a1c00?)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:273 +0xe5

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitFunctionCall(0xc006132780, 0xc006132780?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:152 +0x4a

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x24a1500?, 0xc006132780?}, 0xc003f2c600?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:323 +0x146

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitExpressions({0xc004f1a810, 0x1, 0x1}, 0x0?, 0x0?)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:273 +0xe5

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.visitFunctionCall(0xc006132880, 0xc006132880?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:152 +0x4a

github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model.VisitExpression({0x24a1500?, 0xc006132880?}, 0x24a1500?, 0xc00840dc78)

	/home/runner/work/pulumi/pulumi/pkg/codegen/hcl2/model/visitor.go:323 +0x146

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).rewriteQuotes(0xc0008e3c80, {0x24a1500?, 0xc006132880?})

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program_quotes.go:268 +0x67

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).lowerExpression(0xc0008e3c80, {0x24a1500?, 0xc0006fba80?}, {0x24a1200, 0xc0008e2e40})

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program_expressions.go:34 +0x113

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).genResourceDeclaration(0xc0008e3c80, {0x2479240, 0xc000626150}, 0xc000861b80, 0x1)

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program.go:752 +0x2478

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).genResource(...)

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program.go:925

github.com/pulumi/pulumi/pkg/v3/codegen/python.(*generator).genNode(0xc0008e3c80?, {0x2479240?, 0xc000626150?}, {0x24a0ab8?, 0xc000861b80?})

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program.go:556 +0x45

github.com/pulumi/pulumi/pkg/v3/codegen/python.GenerateProgram(0xc000b65158?)

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program.go:70 +0x13c

github.com/pulumi/pulumi/pkg/v3/codegen/python.GenerateProject({0xc0006466b0, 0x6}, {{0xc000d06960, 0x4}, {{0x0, 0x0}, 0x0}, {0x0, 0x0}, 0x0, ...}, ...)

	/home/runner/work/pulumi/pulumi/pkg/codegen/python/gen_program.go:304 +0x3d

main.(*pythonLanguageHost).GenerateProject(0x0?, {0xc0008117d8?, 0x4105c5?}, 0xc0006fa000)

	/home/runner/work/pulumi/pulumi/sdk/python/cmd/pulumi-language-python/main.go:1052 +0x34b

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler.func1({0x2496ca0, 0xc0008029c0}, {0x1e017e0?, 0xc0006fa000})

	/home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:445 +0x72

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x2496ca0, 0xc0008020c0}, {0x1e017e0, 0xc0006fa000}, 0xc000804120, 0xc0008003f0)

	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x3d0

github.com/pulumi/pulumi/sdk/v3/proto/go._LanguageRuntime_GenerateProject_Handler({0x1e5f500?, 0xc0001c8cc0}, {0x2496ca0, 0xc0008020c0}, 0xc000642000, 0xc00066a700)

	/home/runner/work/pulumi/pulumi/sdk/proto/go/language_grpc.pb.go:447 +0x135

google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001bc780, {0x249ec50, 0xc0001ac000}, 0xc0007f8120, 0xc0006df7a0, 0x33b0990, 0x0)

	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:1360 +0xe15

google.golang.org/grpc.(*Server).handleStream(0xc0001bc780, {0x249ec50, 0xc0001ac000}, 0xc0007f8120, 0x0)

	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:1737 +0x9e7

google.golang.org/grpc.(*Server).serveStreams.func1.1()

	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:982 +0x8d

created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 29

	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.57.0/server.go:980 +0x165

error: could not generate output program: error reading from server: EOF```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20231004--programgen-python--fix-panic-in-python-program-gen-when-rewriting-index-expressions.yaml
+++ b/changelog/pending/20231004--programgen-python--fix-panic-in-python-program-gen-when-rewriting-index-expressions.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen/python
+  description: Fix panic in python program-gen when rewriting index expressions

--- a/pkg/codegen/python/gen_program_quotes.go
+++ b/pkg/codegen/python/gen_program_quotes.go
@@ -68,7 +68,9 @@ func (g *generator) rewriteTraversal(traversal hcl.Traversal, source model.Expre
 
 		if objectKey && isLegalIdentifier(keyVal) {
 			currentTraversal = append(currentTraversal, traverser)
-			currentParts = append(currentParts, parts[i+1])
+			if i < len(traversal)-1 {
+				currentParts = append(currentParts, parts[i+1])
+			}
 			continue
 		}
 

--- a/pkg/codegen/testing/test/program_driver.go
+++ b/pkg/codegen/testing/test/program_driver.go
@@ -379,6 +379,11 @@ var PulumiPulumiProgramTests = []ProgramTest{
 		Directory:   "empty-list-property",
 		Description: "Tests compiling empty list expressions of object properties",
 	},
+	{
+		Directory:   "python-regress-14037",
+		Description: "Regression test for rewriting qoutes in python",
+		Skip:        allProgLanguages.Except("python"),
+	},
 }
 
 var PulumiPulumiYAMLProgramTests = []ProgramTest{

--- a/pkg/codegen/testing/test/testdata/python-regress-14037-pp/python-regress-14037.pp
+++ b/pkg/codegen/testing/test/testdata/python-regress-14037-pp/python-regress-14037.pp
@@ -1,0 +1,14 @@
+data = ["bob", "john", "carl"]
+
+resource "user" "random:index/randomPassword:RandomPassword" {
+  options { range = data }
+  length = 16
+}
+resource "dbUsers" "aws:secretsmanager/secretVersion:SecretVersion" {
+  options { range = data }
+  secretId = "mySecret"
+  secretString = toJSON({
+    username = range.value
+    password = user[range.value].result
+  })
+}

--- a/pkg/codegen/testing/test/testdata/python-regress-14037-pp/python/python-regress-14037.py
+++ b/pkg/codegen/testing/test/testdata/python-regress-14037-pp/python/python-regress-14037.py
@@ -1,0 +1,21 @@
+import pulumi
+import json
+import pulumi_aws as aws
+import pulumi_random as random
+
+data = [
+    "bob",
+    "john",
+    "carl",
+]
+user = []
+for range in [{"key": k, "value": v} for [k, v] in enumerate(data)]:
+    user.append(random.RandomPassword(f"user-{range['key']}", length=16))
+db_users = []
+for range in [{"key": k, "value": v} for [k, v] in enumerate(data)]:
+    db_users.append(aws.secretsmanager.SecretVersion(f"dbUsers-{range['key']}",
+        secret_id="mySecret",
+        secret_string=user[range["value"]].result.apply(lambda result: json.dumps({
+            "username": range["value"],
+            "password": result,
+        }))))

